### PR TITLE
Webhook: Integration test

### PR DESCRIPTION
**Pull request by aandvalenzuela**

### CODE REVIEW

File: test/src/406-ducc-webhook-notifications/main [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+send_delete() {
+    curl -X POST http://webhook-c8vm:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"DELETE_ARTIFACT\", \"event_data\": {\"resources\": [{\"resource_url\": \"registry.cern.ch/unpacked-dev/image:tag\"}]}}" || return 4
+}
+
+send_replication() {
+    curl -X POST http://webhook-c8vm:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"REPLICATION\", \"event_data\": {\"replication\": { \"dest_resource\": {\"endpoint\": \"https://registry.cern.ch\", \"namespace\": \"unpacked-dev\"}, \"successful_artifact\": [{\"type\": \"image\", \"status\": \"Success\", \"name_tag\": \"image:tag\"}]}}}" || return 5
+}
+
+cvmfs_run_test() {
+
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]: Remove empty line

File: test/src/406-ducc-webhook-notifications/main [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    notifications_file="notifications.txt"
+    sudo python3 /root/cvmfs/test/src/406-ducc-webhook-notifications/registry-webhook.py -f $notifications_file -h 188.185.120.234 -p 8080 &
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]: Are you sure this needs to run as `sudo`?

File: test/src/406-ducc-webhook-notifications/main [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    sleep 1
+    pid=$(echo $!)
+
+    yes | touch $notifications_file
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]: This should work without a confirmation, I'd think

File: test/src/406-ducc-webhook-notifications/main [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+send_replication() {
+    curl -X POST http://webhook-c8vm:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"REPLICATION\", \"event_data\": {\"replication\": { \"dest_resource\": {\"endpoint\": \"https://registry.cern.ch\", \"namespace\": \"unpacked-dev\"}, \"successful_artifact\": [{\"type\": \"image\", \"status\": \"Success\", \"name_tag\": \"image:tag\"}]}}}" || return 5
+}
+
+cvmfs_run_test() {
+
+    notifications_file="notifications.txt"
+    sudo python3 /root/cvmfs/test/src/406-ducc-webhook-notifications/registry-webhook.py -f $notifications_file -h 188.185.120.234 -p 8080 &
+
+    sleep 1
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]: Does the test work without sleep? Sleeping is likely to make the test flaky.

File: test/src/406-ducc-webhook-notifications/main [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        echo 'The action not correct'
+        kill $pid
+        return 2
+    fi
+
+    sleep 1
+}
+
+send_push() {
+    curl -X POST http://webhook-c8vm:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"PUSH_ARTIFACT\", \"event_data\": {\"resources\": [{\"resource_url\": \"registry.cern.ch/unpacked-dev/image:tag\"}]}}" || return 3
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]:
```suggestion
    curl -X POST http://localhost:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"PUSH_ARTIFACT\", \"event_data\": {\"resources\": [{\"resource_url\": \"registry.cern.ch/unpacked-dev/image:tag\"}]}}" || return 3
```

File: test/src/406-ducc-webhook-notifications/main [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+send_delete() {
+    curl -X POST http://webhook-c8vm:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"DELETE_ARTIFACT\", \"event_data\": {\"resources\": [{\"resource_url\": \"registry.cern.ch/unpacked-dev/image:tag\"}]}}" || return 4
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]:
```suggestion
    curl -X POST http://localhost:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"DELETE_ARTIFACT\", \"event_data\": {\"resources\": [{\"resource_url\": \"registry.cern.ch/unpacked-dev/image:tag\"}]}}" || return 4
```

File: test/src/406-ducc-webhook-notifications/main [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+send_replication() {
+    curl -X POST http://webhook-c8vm:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"REPLICATION\", \"event_data\": {\"replication\": { \"dest_resource\": {\"endpoint\": \"https://registry.cern.ch\", \"namespace\": \"unpacked-dev\"}, \"successful_artifact\": [{\"type\": \"image\", \"status\": \"Success\", \"name_tag\": \"image:tag\"}]}}}" || return 5
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]:
```suggestion
    curl -X POST http://localhost:8080/webhooks/test-cvmfs-ingestion -H "Content-Type: application/json" -d "{\"type\": \"REPLICATION\", \"event_data\": {\"replication\": { \"dest_resource\": {\"endpoint\": \"https://registry.cern.ch\", \"namespace\": \"unpacked-dev\"}, \"successful_artifact\": [{\"type\": \"image\", \"status\": \"Success\", \"name_tag\": \"image:tag\"}]}}}" || return 5
```

File: test/src/406-ducc-webhook-notifications/main [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    expected_id=$1
+    expected_action=$2
+
+    notification=$( tail -n 1 $notifications_file )
+    id=(${notification//|/ })
+
+    if [ ${id[0]} != ${expected_id} ]
+    then
+        echo 'The id of the action is not correct'
+        kill $pid
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]: Let's use a `trap` to stop the webhook.  You can search for `trap` in other integration tests and use it as a template

File: test/src/406-ducc-webhook-notifications/main [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    fi
+
+    if [ ${id[1]} != ${expected_action} ]
+    then
+        echo 'The action not correct'
+        kill $pid
+        return 2
+    fi
+
+    sleep 1
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]: Let's try to remove the sleep

File: test/src/406-ducc-webhook-notifications/main [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+    send_delete
+    check_id_action '6' 'delete'
+
+    echo "999|push|https://registry.cern.ch/unpacked-dev/image:tag" >> $notifications_file
+
+    send_delete
+    check_id_action '1000' 'delete'
+
+    original_content=$(<$notifications_file)
```
Comment (jblomer) [inline test/src/406-ducc-webhook-notifications/main]: Might be easier to just copy the file to a temporary one and than use `diff` to compare with the rotated.

**Review by jblomer — CHANGES_REQUESTED** [review]
Very good! A few comments inline

**Comment by aandvalenzuela** [comment]
@cernvm-bot cloudtest

**Comment by aandvalenzuela** [comment]
I close this PR to continue in [#2770](https://github.com/cvmfs/cvmfs/pull/2770).

**Pull request closed without merging**